### PR TITLE
take/return in typechecker, etc

### LINKED ISSUE #549: Refactor for take / return in application, refactor typeOf
- Move `Take` and `Return` out of `ErasedAnn`, because the typechecker should not import, to some util file
- Update typechecker interface to use `Take` and `Return` (generic over the prim ty & prim val)
- Make primitive types polymorphic (e.g. pairs in Michelson)
- Update Michelson backend to use `applyProper` ([here](https://github.com/metastatedev/juvix/blob/develop/src/Juvix/Backends/Michelson/Parameterisation.hs#L58))

### PR DESCRIPTION
**Pull request by andy-morris**
Fixes #549 and #555.

- [x] move `Take` and `Return` to `Juvix.Core.Application`
- [x] add type to `Return` constructor
- [x] split michelson parameterisation into `RawPrimVal` (bare instructions, constants) and `PrimVal` (an alias for `Return`—typed, supports partial application)
- [x] move `apply` and `arity` from parameterisation to `CanApply` class (and make `apply` n-ary)
  - (they could also be put in a separate record if there ever needs to be more than one for a type, but that seems unlikely?)
- [x] make application for `PrimVal` use the full `applyProper` stuff
- [x] make typechecker output a `Term primTy (Return _ primVal)`, which adds specific types to the prims and makes them partially applyable later
- [x] fix `Evaluator` for `CanApply`
- [x] fix `InstructionsEff` for wrapped prims
- [x] fix arith circuit backend for parameterisation changes
- [x] fill the non-match-related holes in `FromFrontend`
- [x] fixup tests
- [x] remember to run `make format` for once in my life
- [x] rebase

### CODE REVIEW

**Comment by andy-morris** [comment]
reminder to myself mostly: pushing everything non-michelson related to a later pr

**Comment by cwgoes** [comment]
At the moment I can't build; I receive the following error:

```bash

/home/cwgoes/working/cryptium/juvix/src/Juvix/Core/FromFrontend.hs:304:24-46: error:
    • Couldn't match expected type ‘IR.GlobalUsage’
                  with actual type ‘Usage.NatAndw’
    • In the ‘funUsage’ field of a record
      In the expression:
        IR.Function
          {funName = x, funUsage = fromMaybe Usage.Omega π, funType = _,
           funClauses = _}
      In an equation for ‘f’:
          f = IR.Function
                {funName = x, funUsage = fromMaybe Usage.Omega π, funType = _,
[11 of 13] Compiling Juvix.Core.Pipeline
                 funClauses = _}
[12 of 13] Compiling Juvix.Core
    |
[13 of 13] Compiling Juvix.Pipeline
304 |             funUsage = fromMaybe Usage.Omega π,
    |                        ^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by andy-morris** [comment]
@cwgoes ugh sorry i forgot to stage this last little bit. it should work now? (other than the tests)

**Comment by cwgoes** [comment]
Thanks; I can build now. What do we need to do to update the tests, just use the latest stuff from Core/HR?

**Comment by cwgoes** [comment]
I merged `develop` for the latest `stack.yaml`, now it looks like just a minor change is required:

```bash
juvix             > /home/cwgoes/working/cryptium/juvix/src/Juvix/Pipeline.hs:33:61-85: error:
juvix             >     • Couldn't match type ‘Juvix.FrontendContextualise.Final
juvix             >                              Context.Definition
juvix             >                            -> FromFrontend.Env primTy0 primVal0 [IR.Global primTy0 primVal0]’
juvix             >                      with ‘FromFrontend.Env primTy primVal [IR.Global primTy primVal]’
juvix             >       Expected type: Context.Definition
juvix             >                        (NonEmpty
juvix             >                           (Juvix.FrontendContextualise.InfixPrecedence.Types.FunctionLike
juvix             >                              Juvix.FrontendContextualise.InfixPrecedence.Types.Expression))
juvix             >                        Juvix.FrontendContextualise.InfixPrecedence.Types.Signature
juvix             >                        Juvix.FrontendContextualise.InfixPrecedence.Types.Type
juvix             >                      -> FromFrontend.Env primTy primVal [IR.Global primTy primVal]
juvix             >         Actual type: Symbol
juvix             >                      -> Juvix.FrontendContextualise.Final Context.Definition
juvix             >                      -> FromFrontend.Env primTy0 primVal0 [IR.Global primTy0 primVal0]
juvix             >     • Probable cause: ‘FromFrontend.transformDef’ is applied to too few arguments
juvix             >       In the first argument of ‘Context.traverseContext1’, namely
juvix             >         ‘FromFrontend.transformDef’
juvix             >       In the second argument of ‘($)’, namely
juvix             >         ‘Context.traverseContext1 FromFrontend.transformDef ctx’
juvix             >       In the expression:
juvix             >         FromFrontend.execEnv ctx param
juvix             >           $ Context.traverseContext1 FromFrontend.transformDef ctx
juvix             >     • Relevant bindings include
juvix             >         param :: P.Parameterisation primTy primVal
juvix             >           (bound at src/Juvix/Pipeline.hs:32:19)
juvix             >         contextToCore :: Target.FinalContext
juvix             >                          -> P.Parameterisation primTy primVal
juvix             >                          -> Either FromFrontend.Error [IR.Global primTy primVal]
juvix             >           (bound at src/Juvix/Pipeline.hs:32:1)
juvix             > [13 of 13] Compiling Paths_juvix
juvix             >    |
juvix             > 33 |   FromFrontend.execEnv ctx param $ Context.traverseContext1 FromFrontend.transformDef ctx


```

File: src/Juvix/Core/FromFrontend.hs [hunk 1]
```diff
@@ -236,50 +267,62 @@ transformUsage (FE.Constant (FE.Number (FE.Integer' i)))
...
-transformDef (Ctx.TypeDeclar typ) = _
-transformDef (Ctx.Unknown _) = pure []
-transformDef Ctx.CurrentNameSpace = pure [] -- FIXME ???
+transformDef x (Ctx.Def π sig def _) = do
+  π <- maybe (pure IR.GOmega) usageToGlobal π
+  let f =
+        IR.Function
+          { funName = x,
+            funUsage = π,
+            funType = _, -- TODO pass in answer from transformTypeSig
```
Comment (cwgoes) [inline src/Juvix/Core/FromFrontend.hs:315]: Looks like there are still some holes here?

File: src/Juvix/Core/FromFrontend.hs [hunk 2]
```diff
@@ -236,50 +267,62 @@ transformUsage (FE.Constant (FE.Number (FE.Integer' i)))
...
+            funClauses = _
+          }
+  pure [IR.GFunction f]
+transformDef _ d@(Ctx.Record _ _) = throwFF $ DefUnsupported d
+transformDef _ (Ctx.TypeDeclar typ) = _
```
Comment (cwgoes) [inline src/Juvix/Core/FromFrontend.hs:320]: Also still a hole?

**Review by cwgoes — CHANGES_REQUESTED** [review]
Took a quick pass, generally looks good but haven't tested yet.

cc @mariari can you look at the Michelson backend changes

**Comment by mariari** [comment]
> Took a quick pass, generally looks good but haven't tested yet.
> 
> cc @mariari can you look at the Michelson backend changes

All the Michelson backend changes look reasonable. Most of it besides param are just changing a type's name around.

As for Param itself, the code looks like a straight forward translation to a proper core typeclass over the concept.

File: library/Core/src/Juvix/Core/Application.hs [hunk 3]
```diff
@@ -0,0 +1,32 @@
+-- |
+-- - This module represents the type which will be sent to the
+--   parameterisation
+-- - the =Take= type is what a parameterisation will take coming in
+-- - the =Return= type is what will be handed back to Core to evaluate
+--   and decide on the next steps. If this is a =Left= type checking
```
[the remark this answers is not in the document]
Comment (andy-morris) [reply library/Core/src/Juvix/Core/Application.hs]: That's the comment they had originally, sorry. I've updated it with something more accurate

File: library/Core/src/Juvix/Core/ErasedAnn/Prim.hs [hunk 4]
```diff
@@ -1,39 +1,41 @@
...
+fromAnn _ = Nothing
 
 toAnn :: Take primTy primVal -> Types.AnnTerm primTy primVal
-toAnn (Take usage type' term) = Types.Ann usage type' (Types.Prim term)
+toAnn (App.Take usage type' term) = Types.Ann usage type' $ Types.Prim term
+
+fromPrimType :: P.PrimType primTy -> Types.Type primTy
+fromPrimType (t :| []) = Types.PrimTy t
+fromPrimType (t :| (u : us)) =
+  Types.Pi Usage.Omega (Types.PrimTy t) (fromPrimType (u :| us))
```
[the remark this answers is not in the document]
Comment (andy-morris) [reply library/Core/src/Juvix/Core/ErasedAnn/Prim.hs:41]: I thought we could give the primitives all `ω A → ω B → ... → Z` types and wrap them in the library with the actual usages, but now I'm tihnking about it again, I might have actually got the subtyping the wrong way round. In that case the parameterisation will need to know the usages as well as the arg types.

[the remark this answers is not in the document]
Comment (cwgoes) [reply library/Core/src/Juvix/Core/ErasedAnn/Prim.hs:41]: For now I think we can require that the parameterisation provide usage annotations for all primitive function types.

File: library/Core/src/Juvix/Core/Parameterisation.hs [hunk 5]
```diff
@@ -32,3 +33,44 @@ data Parameterisation primTy primVal
...
+  = ExtraArguments a (NonEmpty a)
+  | InvalidArguments a (NonEmpty a)
+  | Extra e
+  deriving (Eq, Show, Functor)
+
+type ApplyError a = ApplyError' (ApplyErrorExtra a) a
+
+class CanApply a where
+  type ApplyErrorExtra a
+  type ApplyErrorExtra a = Void
```
Comment (mariari) [inline library/Core/src/Juvix/Core/Parameterisation.hs:55]: What does the double type declaration do here

File: src/Juvix/Core/FromFrontend.hs [hunk 6]
```diff
@@ -162,71 +249,83 @@ transformTermHR (FE.Primitive (FE.Prim p)) = do
...
+  throwFF $ UniversesUnimplemented i
 transformTermHR (FE.Parened e) = transformTermHR e
 
+transformSimpleLet :: FE.Let -> Env primTy primVal (HR.Term primTy primVal)
+transformSimpleLet e@(FE.LetGroup name (clause :| []) body) = do
+  let FE.Like args cbody = clause
+  rhs <-
+    toElim (FE.Let e)
+      =<< foldr HR.Lam <$> transformTermHR cbody <*> traverse isVarArg args
+  HR.Let Usage.Omega (NameSymbol.fromSymbol name) rhs <$> transformTermHR body
```
Comment (mariari) [inline src/Juvix/Core/FromFrontend.hs:278]: This seems reasonable modulo the usage question.

The frontend usage is my bad, given we don't have let sig rolled up I don't think

File: src/Juvix/Core/FromFrontend.hs [hunk 7]
```diff
@@ -236,50 +335,192 @@ transformUsage (FE.Constant (FE.Number (FE.Integer' i)))
...
+  clauses <- traverse transformClause def
+  let f =
+        IR.Function
+          { funName = x,
+            funUsage = π,
+            funType = hrToIR typ,
+            funClauses = clauses
+          }
+  pure [IR.GFunction f]
+transformDef _ d@(Ctx.Record _ _) = throwFF $ DefUnimplemented d
```
Comment (mariari) [inline src/Juvix/Core/FromFrontend.hs:405]: Don't we have an issue where any module name `Bar.Foo` will trigger this def undefined?

Do we secure against this anywhere?

File: src/Juvix/Core/Pipeline.hs [hunk 8]
```diff
@@ -1,67 +1,94 @@
...
+
 module Juvix.Core.Pipeline where
 
 import qualified Data.HashMap.Strict as HM
 import qualified Juvix.Backends.Michelson as Michelson
+import qualified Juvix.Core.Application as App
 import qualified Juvix.Core.ErasedAnn as ErasedAnn
+import qualified Juvix.Core.ErasedAnn.Prim as Prim
 import qualified Juvix.Core.Erasure as Erasure
+import qualified Juvix.Core.Erasure.Types as Erasure
```
Comment (mariari) [inline src/Juvix/Core/Pipeline.hs:11]: Seems unused? 

At least that is what my editor is telling me

File: src/Juvix/Core/Pipeline.hs [hunk 9]
```diff
@@ -108,11 +135,44 @@ coreToMichelsonContract term usage ty = do
...
+  ( HasWriter "log" [Types.PipelineLog primTy primVal] m,
+    HasReader "parameterisation" (Types.Parameterisation primTy primVal) m,
+    HasThrow "error" (Types.PipelineError primTy primVal compErr) m,
+    HasReader "globals" (IR.GlobalsT primTy primVal) m,
+    Types.CanApply (Typed.TypedPrim primTy primVal),
+    Eq primTy,
+    Eq primVal,
+    Show primTy,
+    Show primVal,
+    Show compErr
```
Comment (mariari) [inline src/Juvix/Core/Pipeline.hs:148]: We really should start maing alias for these

**Review by mariari — COMMENTED** [review]
Overall seems fine

File: library/Core/src/Juvix/Core/EAC/Parser.hs [hunk 10]
```diff
@@ -60,17 +62,22 @@ langaugeDef =
...
 
 identifier :: Stream s m Char => ParsecT s u m String
 identifier = T.identifier lexer
 
+qualIdentifier :: Stream s m Char => ParsecT s u m (NonEmpty String)
+qualIdentifier =
+  -- TODO maybe a different separator so that "λ x.x" looks less misleading?
+  NonEmpty.fromList <$> identifier `sepBy1` reservedOp "."
+
 reserved :: Stream s m Char => String -> ParsecT s u m ()
```
[the remark this answers is not in the document]
Comment (andy-morris) [reply library/Core/src/Juvix/Core/EAC/Parser.hs:81]: Sure. I just didn't want to change the syntax without telling anyone. But we can definitely do that if there's general agreement.

File: library/Core/src/Juvix/Core/Parameterisation.hs (hunk 5 above)
Comment (andy-morris) [reply library/Core/src/Juvix/Core/Parameterisation.hs:55]: The one with an RHS is a default, so if you don't have any special type errors you can just omit it entirely

File: library/Core/src/Juvix/Core/EAC/Parser.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (cwgoes) [reply library/Core/src/Juvix/Core/EAC/Parser.hs:81]: I'm fine with that change.

File: library/Core/src/Juvix/Core/ErasedAnn/Conversion.hs [hunk 11]
```diff
@@ -4,62 +4,61 @@ import Data.List ((\\))
...
-          pure (Ann usage ty' (AppM f' (a' <> [a])))
-        _ ->
-          pure (Ann usage ty' (AppM f [a]))
+      ty' = convertType ty
+   in case term of
+        E.Var sym _ -> Ann usage ty' (Var sym)
+        E.Prim p _ -> Ann usage ty' (Prim p)
+        E.Let sym bind body (bindTy, _) ->
+          -- Calculate captures.
+          let captures = Erased.free (Erased.Lam sym (E.eraseAnn body))
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply library/Core/src/Juvix/Core/ErasedAnn/Conversion.hs:23]: Yeah I think it was left over from a previous version of the function

File: library/Core/src/Juvix/Core/ErasedAnn/Prim.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cwgoes) [reply library/Core/src/Juvix/Core/ErasedAnn/Prim.hs:41]: I presume this has been split to https://github.com/metastatedev/juvix/issues/578

File: src/Juvix/Core/FromFrontend.hs (hunk 6 above)
Comment (cwgoes) [reply src/Juvix/Core/FromFrontend.hs:278]: Yeah I guess we'll need an explicit usage here for now, that's fine.

**Review by cwgoes — APPROVED** [review]
ACK modulo minor comments (and testing, to come after merge) :tada:

File: library/Core/src/Juvix/Core/EAC/Parser.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (mariari) [reply library/Core/src/Juvix/Core/EAC/Parser.hs:81]: Once we change it, we'd need to update /doc/Bohm/ebnf.org/ and the generated diagram from the grammar IIRC.

**Comment by cwgoes** [comment]
@andy-morris Are there other things you want to do here (it's still marked "draft")?

File: library/Core/src/Juvix/Core/EAC/Parser.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (cwgoes) [reply library/Core/src/Juvix/Core/EAC/Parser.hs:81]: Split to https://github.com/metastatedev/juvix/issues/580

File: src/Juvix/Core/FromFrontend.hs (hunk 6 above)
Comment (cwgoes) [reply src/Juvix/Core/FromFrontend.hs:278]: Split to new issue https://github.com/metastatedev/juvix/issues/581

File: src/Juvix/Core/FromFrontend.hs (hunk 7 above)
Comment (cwgoes) [reply src/Juvix/Core/FromFrontend.hs]: (since updated)

**Pull request merged**